Worker: Perf, avoid dynamic memory allocation in TransportTuple hash

jmillan opened the pull request:
For every packet arriving from network we are generating a tuple, and upon creation we create its hash.

Before this PR we were using a `std::vector` which we later fill it with the tuple info. This generates dynamic memory allocation which must later be released.

This PR makes use of `std::array` with a fixed length (we know it in advance), so we store the info only in the stack.

This only change improves CPU performance in a little more than 1%.

Before `TransportTuple::GenerateHash()` took around 1.4%:
<img width="775" height="85" alt="Screenshot 2025-11-26 at 15 28 25" src="https://github.com/user-attachments/assets/ac7e5e40-0075-4df4-967f-a545415ae2fb" />

After  `TransportTuple::GenerateHash()` takes around 0.2%:
<img width="625" height="20" alt="Screenshot 2025-11-26 at 15 17 35" src="https://github.com/user-attachments/assets/397da425-7b0b-4fb5-a1ae-0c6a8bb2abdb" />

jmillan committed:
Worker: Perf, avoid dynamic memory allocation in TransportTuple hash

For every packet arriving from network we are generating a tuple, and upon creation
we create its hash.

Before this PR we were using a std::vector which we later fill it with the tuple info.
This generates dynamic memory allocation which must later be released.

This PR makes use of std::array with a fixed length (we know it in advance).

This only changes improves CPU performance in 1%.

ibc commented on worker/src/RTC/TransportTuple.cpp:
```diff
@@ -153,7 +153,10 @@ namespace RTC
-		std::vector<uint8_t> buffer;
+		// Maximum buffer length for two IPv6 addresses and ports plus protocol.
+		static constexpr size_t BufferSize = ((16 + 2) * 2) + 1;
+		std::array<uint8_t, BufferSize> buffer{};
```
We should include std::array at the top of this file

ibc replied:
👆🏻👆🏻👆🏻

ibc commented on worker/src/RTC/TransportTuple.cpp:
```diff
@@ -163,26 +166,26 @@ namespace RTC
-				buffer.push_back((port >> 8) & 0xFF);
-				buffer.push_back(port & 0xFF);
+				it    = std::copy(ip, ip + 4, it);
+				*it++ = (port >> 8) & 0xFF;
```
Wow I don't really know what this `*it++ =` does

jmillan replied:
Assign the value of the right operand to `it` and increment it.

nazar-pc replied:
Why not write separate operations on separate lines, so it is easier to read?

jmillan replied:
If `it` if for example `buffer.begin()`, this is index 0, then it will assign `(port >> 8) & 0xFF` to the first position of the array and `it` will point now to the position immediately after the written data (hence the ++), this is, index 1.

nazar-pc replied:
The fact that two people ask what this code does is a pretty strong indication that less fancy code that takes 3 lines instead of one would be much more maintainable.

jmillan replied:
Yes, it definitely makes sense to make it a bit more verbose.

jmillan replied:
> The fact that two people ask what this code does is a pretty strong indication that less fancy code that takes 3 lines instead of one would be much more maintainable.

BTW I responded to Iñaki before reading your comment.

jmillan committed:
include std::array

jmillan commented:
Windows failing.

ibc commented:
> * mediasoup-worker / ci (windows-2025, cl, cl, Release, false, false, false, false, false, 24) (pull_request)

Here:

```
../../../src/RTC/TransportTuple.cpp(159): error C3535: cannot deduce type for 'auto *' from 'std::_Array_iterator<_Ty,37>'
        with
        [
            _Ty=uint8_t
        ]
../../../src/RTC/TransportTuple.cpp(159): error C2440: 'initializing': cannot convert from 'std::_Array_iterator<_Ty,37>' to '<error type> *'
        with
        [
            _Ty=uint8_t
        ]
../../../src/RTC/TransportTuple.cpp(159): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
[1296/1386] Compiling C++ object mediasoup-worker.exe.p/src_RTC_SimulcastConsumer.cpp.obj
```

jmillan committed:
make array edition more verbose

Split in two the write and increment operation of the iterator.

Also specifiy type for `buffer.begin()` as Windows cannot deduce it from auto.

jmillan commented:
@ibc, @nazar-pc I split the write and increment of the iterator in two lines. I hope this makes it more readable.

jmillan committed:
use the old good array

std::array does not provide the number of added elements, just the total size.

We need it in order to call GenerateFnv1aHash() with the proper size. Previous to this
commit we were passing the whole array capacity rather than the real size, meaning
extra length for IPv4 tuples (it does not hurt, but it's not correct either).

Since we need an incremental index anyway there is no reason to keep using std::array.

ibc commented:
Please add the #include corresponding to std::memcpy() in the file

jmillan committed:
include file for std::memcpy
